### PR TITLE
Dynamically load postgres extensions

### PR DESCRIPTION
**Pull request by lubennikovaav**
## Summary of changes

If extension file is not found - request it from compute_ctl

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Review by awestover — APPROVED** [review]
I added my download files functionality

File: compute_tools/src/bin/compute_ctl.rs [hunk 1]
```diff
@@ -184,6 +184,8 @@ fn main() -> Result<()> {
     let _http_handle =
         launch_http_server(http_port, &compute).expect("cannot launch http endpoint thread");
 
+    let extension_server_port: u16 = http_port;
```
Comment (skyzh) [inline compute_tools/src/bin/compute_ctl.rs:187]: I prefer to put it inside compute spec?

File: libs/remote_storage/src/s3_bucket.rs [hunk 2]
```diff
@@ -332,6 +332,53 @@ impl RemoteStorage for S3Bucket {
         Ok(document_keys)
     }
 
+    async fn list_files(
+        &self,
+        folder: Option<&RemotePath>
+    ) -> anyhow::Result<Vec<RemotePath>>{
+        // TODO: should we use DownloadError error type instead of anyhow::Error?
```
Comment (skyzh) [inline libs/remote_storage/src/s3_bucket.rs:339]: should we do the path append `/` thing here?

File: libs/remote_storage/src/s3_bucket.rs [hunk 3]
```diff
@@ -332,6 +332,53 @@ impl RemoteStorage for S3Bucket {
...
+                .await
+                .map_err(|e| {
+                    metrics::inc_list_objects_fail();
+                    e
+                })
+                .context("Failed to list files in S3 bucket")?;
+        
+            for object in response.contents().unwrap_or_default() {
+                let object_path = object.key().unwrap();
+                println!("{:?}", object_path);
```
Comment (skyzh) [inline libs/remote_storage/src/s3_bucket.rs:370]: rm printlns?

**Review by skyzh — COMMENTED** [review]
Generally LGTM with the S3 list file part. It would be better if there could be some integration tests (as how we test the `list_prefix` function). The S3/local_fs list file can also be a separate PR so that we can get it merged quickly :)

File: libs/remote_storage/src/s3_bucket.rs (hunk 2 above)
Comment (awestover) [reply libs/remote_storage/src/s3_bucket.rs:339]: For me, if you put a `/` then it doesn't work.

File: libs/remote_storage/src/s3_bucket.rs (hunk 3 above)
Comment (awestover) [reply libs/remote_storage/src/s3_bucket.rs:370]: yes

File: compute_tools/src/bin/compute_ctl.rs (hunk 1 above)
Comment (awestover) [reply compute_tools/src/bin/compute_ctl.rs:215]: we are going to try to start loading extensions before the spec arrives

File: .gitignore [hunk 4]
```diff
@@ -18,3 +18,5 @@ test_output/
 *.o
 *.so
 *.Po
+
+alek/
```
Comment (vadim2404) [inline .gitignore:22]: hopefully we don't need in our repo

Comment (awestover) [reply .gitignore:22]: :) just for testing purposes

File: test_ext.control
Comment (lubennikovaav) [inline test_ext.control:1]: it's better to move it to some subdir in test_runner

File: test_runner/regress/test_download_extensions.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            log.info(cur.fetchall())
+
+            # the real test query: check that test_ext is present
+            cur.execute("SELECT * FROM pg_available_extensions")
+            all_extensions = [x[0] for x in cur.fetchall()]
+            log.info(all_extensions)
+            assert "test_ext" in all_extensions
+
+    endpoint.stop()
+    env.pageserver.http_client().tenant_detach(tenant)
```
Comment (lubennikovaav) [inline test_runner/regress/test_download_extensions.py:83]: This is not needed, pytest will do the cleanup on the test end.

File: compute_tools/src/extension_server.rs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+use serde_json::{self, Value};
+use std::fs::File;
+use std::io::{BufWriter, Write};
+use std::num::{NonZeroU32, NonZeroUsize};
+use std::path::Path;
+use std::str;
+use tokio::io::AsyncReadExt;
+use tracing::info;
+
+fn get_pg_config(argument: &str, pgbin: &str) -> String {
```
Comment (lubennikovaav) [inline compute_tools/src/extension_server.rs:12]: add some comment here about path assumptions and pgbin arg

**Review by awestover — COMMENTED** [review]
looks fine as long as long as the tests pass

**Comment by awestover** [comment]
Note: this PR does not contain all the extension features
In
https://github.com/neondatabase/neon/pull/4555
I plan to add:
```
- add more tests
  -- multiple extension.control files to download
  -- does CREATE EXTENSION actually call extension_server?
- Handle not just Shared Extensions but also other types of extensions
- libs/remote_storage/src/s3_bucket.rs // TODO: if bucket prefix is empty, folder is prefixed with a "/" I think. Is this desired?
- add env variable to github of EXT_REMOTE_STORAGE_S3_BUCKET
- add tests for extension downloading with real S3 storage
```

File: compute_tools/src/http/api.rs [hunk 7]
```diff
@@ -121,8 +123,36 @@ async fn routes(req: Request<Body>, compute: &Arc<ComputeNode>) -> Response<Body
...
+            match extension_server::download_extension(
+                &compute.ext_remote_storage,
+                ExtensionType::Shared,
+                &compute.pgbin,
+            )
+            .await
+            {
+                Ok(_) => Response::new(Body::from("OK")),
+                Err(e) => {
+                    error!("download_file failed: {}", e);
```
Comment (ololobus) [inline compute_tools/src/http/api.rs:146]: NIT: It says `download_file failed`, but I don't see `download_file` in the whole PR

Comment (awestover) [reply compute_tools/src/http/api.rs]: Yeah, it should say `download_extension`. Updated it now.

File: compute_tools/src/extension_server.rs [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+fn get_pg_version(pgbin: &str) -> String {
+    // pg_config --version returns a (platform specific) human readable string
+    // such as "PostgreSQL 15.4". We parse this to v14/v15
+    let human_version = get_pg_config("--version", pgbin);
+    if human_version.contains("v15") {
+        return "v15".to_string();
+    }
+    "v14".to_string()
```
Comment (lubennikovaav) [inline compute_tools/src/extension_server.rs:33]: there will be more supported versions in the future

let's change it to 

```
 else if human_version.contains("v14") {
 return "v14".to_string();
}
 bail!("Unsupported postgres version {human_version}");
```

Comment (awestover) [reply compute_tools/src/extension_server.rs:33]: ok!

**Comment by lubennikovaav** [comment]
closed in favor of rebased version in https://github.com/neondatabase/neon/pull/4592

**Pull request closed without merging**